### PR TITLE
Upgrade Cloudwalk contracts to the 0.8.10 Solidity version.

### PR DESCRIPTION
**Pull request by igorsenych-cw**
# Description
The original Compound smart contracts have been updated to the upstream, which involves using the 0.8.10 Solidity version. Within this PR, all Cloudwalk smart contracts have been updated to the 0.8.10 Solidity version.

## Type of change
- [ ] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: tests/Tokens/setComptrollerTest.js [hunk 1]
```diff
@@ -16,21 +16,22 @@ describe('CToken', function () {
 
   describe('_setComptroller', () => {
     it("should fail if called by non-admin", async () => {
-      expect(
-        await send(cToken, '_setComptroller', [newComptroller._address], { from: accounts[0] })
-      ).toHaveTokenFailure('UNAUTHORIZED', 'SET_COMPTROLLER_OWNER_CHECK');
+      await expect(send(cToken, '_setComptroller', [newComptroller._address], { from: accounts[0] }))
```
Comment (cwteles) [inline tests/Tokens/setComptrollerTest.js:19]: Codacy has a fix for the issue: [Replace `send(cToken,·'_setComptroller',·[newComptroller._address],·{·from:·accounts[0]·})` with `⏎········send(cToken,·"_setComptroller",·[newComptroller._address],·{⏎··········from:·accounts[0],⏎········}`](https://app.codacy.com/gh/cloudwalk/compound-protocol/pullRequest?prid=9803552)
```suggestion
      await expect(
```

File: tests/Errors.js [hunk 2]
```diff
@@ -3,37 +3,56 @@
...
 }
 
 function parse(reporter) {
-  const ErrorInv = reporter.body.find(k => k.name == 'Error').members;
-  const FailureInfoInv = reporter.body.find(k => k.name == 'FailureInfo').members;
-  const Error = invert(ErrorInv);
-  const FailureInfo = invert(FailureInfoInv);
-  return {Error, FailureInfo, ErrorInv, FailureInfoInv};
+  const Error = (reporter.subNodes.find(k => k.name == 'Error') || { members: [] }).members.reduce((obj, i, idx) => {
+    obj[i.name] = `${idx}`
```
Comment (cwteles) [inline tests/Errors.js:24]: Codacy has a fix for the issue: [Insert `;`](https://app.codacy.com/gh/cloudwalk/compound-protocol/pullRequest?prid=9803552)
```suggestion
    obj[i.name] = `${idx}`;
```

File: tests/Tokens/setInterestRateModelTest.js [hunk 3]
```diff
@@ -22,38 +22,35 @@ describe('CToken', function () {
...
+      await expect(send(cToken, 'harnessSetInterestRateModelFresh', [badModel._address]))
+        .rejects.toRevert("revert marker method returned false");
       expect(await call(cToken, 'interestRateModel')).toEqual(oldModel._address);
     });
 
     it("accepts new valid interest rate model", async () => {
-      expect(
-        await send(cToken, 'harnessSetInterestRateModelFresh', [newModel._address])
-      ).toSucceed();
+      expect(await send(cToken, 'harnessSetInterestRateModelFresh', [newModel._address]))
```
Comment (cwteles) [inline tests/Tokens/setInterestRateModelTest.js:52]: Codacy has a fix for the issue: [Replace `await·send(cToken,·'harnessSetInterestRateModelFresh',·[newModel._address]))` with `⏎········await·send(cToken,·"harnessSetInterestRateModelFresh",·[⏎··········newModel._address,`](https://app.codacy.com/gh/cloudwalk/compound-protocol/pullRequest?prid=9803552)
```suggestion
      expect(
```

File: scenario/src/ErrorReporter.ts [hunk 4]
```diff
@@ -18,6 +21,10 @@ class NoErrorReporterType implements ErrorReporter {
   getDetail(error: any, detail: number): string {
     return detail.toString();
   }
+
+  getEncodedCustomError(errorName: string, args: unknown[]): string | null {
+    return null
```
Comment (cwteles) [inline scenario/src/ErrorReporter.ts:26]: Codacy has a fix for the issue: [Insert `;`](https://app.codacy.com/gh/cloudwalk/compound-protocol/pullRequest?prid=9803552)
```suggestion
    return null;
```

File: tests/Errors.js [hunk 5]
```diff
@@ -3,37 +3,56 @@
...
+    return obj
+  }, {});
+  const FailureInfo = (reporter.subNodes.find(k => k.name == 'FailureInfo')  || { members: [] }).members.reduce((obj, i, idx) => {
+    obj[i.name] = `${idx}`
+    return obj
+  }, {});
+  const CustomErrors = reporter.subNodes.filter(k => k.type === 'CustomErrorDefinition').reduce((obj, i) => {
+    obj[i.name] = {
```
Comment (cwteles) [inline tests/Errors.js:32]: Codacy has a fix for the issue: [Insert `··`](https://app.codacy.com/gh/cloudwalk/compound-protocol/pullRequest?prid=9803552)
```suggestion
      obj[i.name] = {
```

File: tests/Errors.js [hunk 6]
```diff
@@ -3,37 +3,56 @@
...
  */
-
+const fs = require('fs')
 const path = require('path');
-const solparse = require('solparse');
+const solparse = require('@solidity-parser/parser');
 
 const errorReporterPath = path.join(__dirname, '..', 'contracts', 'ErrorReporter.sol');
-const contents = solparse.parseFile(errorReporterPath);
+const errorReporterContest = fs.readFileSync(errorReporterPath)
```
Comment (cwteles) [inline tests/Errors.js:11]: Codacy has a fix for the issue: [Insert `;`](https://app.codacy.com/gh/cloudwalk/compound-protocol/pullRequest?prid=9803552)
```suggestion
const errorReporterContest = fs.readFileSync(errorReporterPath);
```

File: tests/Tokens/setInterestRateModelTest.js [hunk 7]
```diff
@@ -22,38 +22,35 @@ describe('CToken', function () {
...
+      await expect(send(cToken, 'harnessSetInterestRateModelFresh', [newModel._address]))
+        .rejects.toRevertWithCustomError('SetInterestRateModelFreshCheck');
       expect(await call(cToken, 'interestRateModel')).toEqual(oldModel._address);
     });
 
     it("reverts if passed a contract that doesn't implement isInterestRateModel", async () => {
-      await expect(
-        send(cToken, 'harnessSetInterestRateModelFresh', [cToken.underlying._address])
-      ).rejects.toRevert();
+      await expect(send(cToken, 'harnessSetInterestRateModelFresh', [cToken.underlying._address]))
```
Comment (cwteles) [inline tests/Tokens/setInterestRateModelTest.js:38]: Codacy has a fix for the issue: [Replace `send(cToken,·'harnessSetInterestRateModelFresh',·[cToken.underlying._address]))` with `⏎········send(cToken,·"harnessSetInterestRateModelFresh",·[⏎··········cToken.underlying._address,`](https://app.codacy.com/gh/cloudwalk/compound-protocol/pullRequest?prid=9803552)
```suggestion
      await expect(
```

File: tests/Tokens/setInterestRateModelTest.js [hunk 8]
```diff
@@ -22,38 +22,35 @@ describe('CToken', function () {
...
+        .rejects.toRevert();
```
Comment (cwteles) [inline tests/Tokens/setInterestRateModelTest.js:39]: Codacy has a fix for the issue: [Insert `])⏎······)`](https://app.codacy.com/gh/cloudwalk/compound-protocol/pullRequest?prid=9803552)
```suggestion
        ])
```

File: tests/Tokens/setInterestRateModelTest.js [hunk 9]
```diff
@@ -22,38 +22,35 @@ describe('CToken', function () {
...
+        .rejects.toRevert("revert marker method returned false");
       expect(await call(cToken, 'interestRateModel')).toEqual(oldModel._address);
     });
 
     it("accepts new valid interest rate model", async () => {
-      expect(
-        await send(cToken, 'harnessSetInterestRateModelFresh', [newModel._address])
-      ).toSucceed();
+      expect(await send(cToken, 'harnessSetInterestRateModelFresh', [newModel._address]))
+        .toSucceed();
```
Comment (cwteles) [inline tests/Tokens/setInterestRateModelTest.js:53]: Codacy has a fix for the issue: [Insert `])⏎······)`](https://app.codacy.com/gh/cloudwalk/compound-protocol/pullRequest?prid=9803552)
```suggestion
        ])
```

File: scenario/src/ErrorReporter.ts [hunk 10]
```diff
@@ -1,9 +1,12 @@
+
+import { utils } from 'ethers';
 import {ComptrollerErr, TokenErr} from './ErrorReporterConstants';
 
 export interface ErrorReporter {
   getError(error: any): string | null
   getInfo(info: any): string | null
   getDetail(error: any, detail: number): string
+  getEncodedCustomError(errorName: string, args: unknown[]): string | null
```
Comment (cwteles) [inline scenario/src/ErrorReporter.ts:9]: Codacy has a fix for the issue: [Insert `;`](https://app.codacy.com/gh/cloudwalk/compound-protocol/pullRequest?prid=9803552)
```suggestion
  getEncodedCustomError(errorName: string, args: unknown[]): string | null;
```

File: scenario/src/ErrorReporter.ts [hunk 11]
```diff
@@ -49,6 +56,14 @@ class CTokenErrorReporterType implements ErrorReporter {
 
     return detail.toString();
   }
+
+  getEncodedCustomError(errorName: string, args: unknown[]): string | null {
+    try {
+      return TokenErr.CustomErrors.encodeErrorResult(errorName, args)
```
Comment (cwteles) [inline scenario/src/ErrorReporter.ts:62]: Codacy has a fix for the issue: [Insert `;`](https://app.codacy.com/gh/cloudwalk/compound-protocol/pullRequest?prid=9803552)
```suggestion
      return TokenErr.CustomErrors.encodeErrorResult(errorName, args);
```

File: tests/Errors.js [hunk 12]
```diff
@@ -3,37 +3,56 @@
...
-const CarefulMath = solparse.parseFile(carefulMathPath).body.find(k => k.type === 'ContractStatement');
-const MathErrorInv = CarefulMath.body.find(k => k.name == 'MathError').members;
+// const carefulMathPath = path.join(__dirname, '..', 'contracts', 'CarefulMath.sol');
+// const CarefulMath = solparse.parseFile(carefulMathPath).body.find(k => k.type === 'ContractStatement');
+const MathErrorInv = {
+  0: "NO_ERROR",
+  1: "DIVISION_BY_ZERO",
+  2: "INTEGER_OVERFLOW",
+  3: "INTEGER_UNDERFLOW",
+} // CarefulMath.body.find(k => k.name == 'MathError').members;
```
Comment (cwteles) [inline tests/Errors.js:51]: Codacy has a fix for the issue: [Insert `;`](https://app.codacy.com/gh/cloudwalk/compound-protocol/pullRequest?prid=9803552)
```suggestion
}; // CarefulMath.body.find(k => k.name == 'MathError').members;
```

File: tests/Tokens/setInterestRateModelTest.js [hunk 13]
```diff
@@ -22,38 +22,35 @@ describe('CToken', function () {
     });
 
     it("fails if called by non-admin", async () => {
-      expect(
-        await send(cToken, 'harnessSetInterestRateModelFresh', [newModel._address], {from: accounts[0]})
-      ).toHaveTokenFailure('UNAUTHORIZED', 'SET_INTEREST_RATE_MODEL_OWNER_CHECK');
+      await expect(send(cToken, 'harnessSetInterestRateModelFresh', [newModel._address], {from: accounts[0]}))
```
Comment (cwteles) [inline tests/Tokens/setInterestRateModelTest.js:25]: Codacy has a fix for the issue: [Replace `send(cToken,·'harnessSetInterestRateModelFresh',·[newModel._address],·{from:·accounts[0]}))` with `⏎········send(cToken,·"harnessSetInterestRateModelFresh",·[newModel._address],·{⏎··········from:·accounts[0],`](https://app.codacy.com/gh/cloudwalk/compound-protocol/pullRequest?prid=9803552)
```suggestion
      await expect(
```

File: tests/Tokens/setInterestRateModelTest.js [hunk 14]
```diff
@@ -22,38 +22,35 @@ describe('CToken', function () {
...
+        .rejects.toRevertWithCustomError('SetInterestRateModelOwnerCheck');
       expect(await call(cToken, 'interestRateModel')).toEqual(oldModel._address);
     });
 
     it("fails if market not fresh", async () => {
       expect(await send(cToken, 'harnessFastForward', [5])).toSucceed();
-      expect(
-        await send(cToken, 'harnessSetInterestRateModelFresh', [newModel._address])
-      ).toHaveTokenFailure('MARKET_NOT_FRESH', 'SET_INTEREST_RATE_MODEL_FRESH_CHECK');
+      await expect(send(cToken, 'harnessSetInterestRateModelFresh', [newModel._address]))
```
Comment (cwteles) [inline tests/Tokens/setInterestRateModelTest.js:32]: Codacy has a fix for the issue: [Replace `send(cToken,·'harnessSetInterestRateModelFresh',·[newModel._address])` with `⏎········send(cToken,·"harnessSetInterestRateModelFresh",·[newModel._address]`](https://app.codacy.com/gh/cloudwalk/compound-protocol/pullRequest?prid=9803552)
```suggestion
      await expect(
```

**Review by EvgeniiZaitsevCW — APPROVED** [review]
Remarks in this review are related to the Compound original code, so they can be skipped.

File: contracts/Cloudwalk/CErc20.sol [hunk 15]
```diff
@@ -238,25 +246,25 @@ contract CErc20 is CToken, CErc20Interface {
         require(success, "TOKEN_TRANSFER_FROM_FAILED");
 
         // Calculate the amount that was *actually* transferred
-        uint balanceAfter = EIP20Interface(underlying).balanceOf(to);
-        require(balanceAfter >= balanceBefore, "TOKEN_TRANSFER_FROM_OVERFLOW");
+        uint balanceAfter = EIP20Interface(underlying_).balanceOf(to);
         return balanceAfter - balanceBefore;   // underflow already checked above, just subtract
```
[the remark this answers is not in the document]
Comment (igorsenych-cw) [reply contracts/Cloudwalk/CErc20.sol:250]: Agree with the comment, but we want to stay as close as possible to the original version of Compound smart contracts.

File: contracts/Cloudwalk/CToken.sol [hunk 16]
```diff
@@ -37,15 +38,15 @@ contract CToken is CTokenInterface, Exponential, TokenErrorReporter {
...
+        require(err == NO_ERROR, "setting comptroller failed");
 
         // Initialize block number and borrow index (block number mocks depend on comptroller being set)
         accrualBlockNumber = getBlockNumber();
         borrowIndex = mantissaOne;
 
         // Set the interest rate model (depends on block number / borrow index)
         err = _setInterestRateModelFresh(interestRateModel_);
-        require(err == uint(Error.NO_ERROR), "setting interest rate model failed");
+        require(err == NO_ERROR, "setting interest rate model failed");
```
[the remark this answers is not in the document]
Comment (igorsenych-cw) [reply contracts/Cloudwalk/CToken.sol:49]: Agree with the comment, but we want to stay as close as possible to the original version of Compound smart contracts.

File: contracts/Cloudwalk/CToken.sol [hunk 17]
```diff
@@ -268,55 +240,39 @@ contract CToken is CTokenInterface, Exponential, TokenErrorReporter {
...
-        require(err == MathError.NO_ERROR, "borrowBalanceStored: borrowBalanceStoredInternal failed");
-        return result;
+    function borrowBalanceStored(address account) override public view returns (uint) {
+        return borrowBalanceStoredInternal(account);
     }
 
     /**
      * @notice Return the borrow balance of account based on stored data
      * @param account The address whose balance should be calculated
      * @return (error code, the calculated balance or 0 if error code is non-zero)
```
[the remark this answers is not in the document]
Comment (igorsenych-cw) [reply contracts/Cloudwalk/CToken.sol:250]: Agree with the comment, but we want to stay as close as possible to the original version of Compound smart contracts.

File: contracts/Cloudwalk/CToken.sol [hunk 18]
```diff
@@ -143,19 +128,19 @@ contract CToken is CTokenInterface, Exponential, TokenErrorReporter {
...
+        return transferTokens(msg.sender, src, dst, amount) == NO_ERROR;
     }
 
     /**
      * @notice Approve `spender` to transfer up to `amount` from `src`
      * @dev This will overwrite the approval amount for `spender`
      *  and is subject to issues noted [here](https://eips.ethereum.org/EIPS/eip-20#approve)
      * @param spender The address of the account which may transfer tokens
-     * @param amount The number of tokens that are approved (-1 means infinite)
+     * @param amount The number of tokens that are approved (uint256.max means infinite)
```
[the remark this answers is not in the document]
Comment (igorsenych-cw) [reply contracts/Cloudwalk/CToken.sol:140]: Agree with the comment, but we want to stay as close as possible to the original version of Compound smart contracts.

File: contracts/Cloudwalk/CToken.sol [hunk 19]
```diff
@@ -37,15 +38,15 @@ contract CToken is CTokenInterface, Exponential, TokenErrorReporter {
 
         // Set the comptroller
         uint err = _setComptroller(comptroller_);
-        require(err == uint(Error.NO_ERROR), "setting comptroller failed");
+        require(err == NO_ERROR, "setting comptroller failed");
```
[the remark this answers is not in the document]
Comment (igorsenych-cw) [reply contracts/Cloudwalk/CToken.sol:41]: Agree with the comment, but we want to stay as close as possible to the original version of Compound smart contracts.

File: contracts/Cloudwalk/CErc20.sol [hunk 20]
```diff
@@ -150,9 +155,11 @@ contract CErc20 is CToken, CErc20Interface {
      *      Note: This wrapper safely handles non-standard ERC-20 tokens that do not return a value.
      *            See here: https://medium.com/coinmonks/missing-return-value-bug-at-least-130-tokens-affected-d67bf08521ca
      */
-    function doTransferIn(address from, uint amount) internal returns (uint) {
-        EIP20NonStandardInterface token = EIP20NonStandardInterface(underlying);
-        uint balanceBefore = EIP20Interface(underlying).balanceOf(address(this));
+    function doTransferIn(address from, uint amount) virtual override internal returns (uint) {
+        // Read from storage once
+        address underlying_ = underlying;
```
Comment (arthur-cw) [inline contracts/Cloudwalk/CErc20.sol:160]: why do we need to do this attribution to a new variable>

File: contracts/Cloudwalk/CErc20.sol [hunk 21]
```diff
@@ -216,9 +222,11 @@ contract CErc20 is CToken, CErc20Interface {
      *      Note: This wrapper safely handles non-standard ERC-20 tokens that do not return a value.
      *            See here: https://medium.com/coinmonks/missing-return-value-bug-at-least-130-tokens-affected-d67bf08521ca
      */
-    function doTransferFrom(address from, address to, uint amount) internal returns (uint) {
-        EIP20NonStandardInterface token = EIP20NonStandardInterface(underlying);
-        uint balanceBefore = EIP20Interface(underlying).balanceOf(to);
+    function doTransferFrom(address from, address to, uint amount) virtual override internal returns (uint) {
+        // Read from storage once
+        address underlying_ = underlying;
```
Comment (arthur-cw) [inline contracts/Cloudwalk/CErc20.sol:227]: same question as above

File: contracts/Cloudwalk/CTokenInterfaces.sol [hunk 22]
```diff
@@ -250,43 +245,43 @@ contract CTokenInterface is CTokenStorage {
...
+    function borrowRatePerBlock() virtual external view returns (uint);
+    function supplyRatePerBlock() virtual external view returns (uint);
+    function totalBorrowsCurrent() virtual external returns (uint);
+    function borrowBalanceCurrent(address account) virtual external returns (uint);
+    function borrowBalanceStored(address account) virtual external view returns (uint);
+    function exchangeRateCurrent() virtual external returns (uint);
+    function exchangeRateStored() virtual external view returns (uint);
+    function getCash() virtual external view returns (uint);
+    function accrueInterest() virtual external returns (uint);
+    function seize(address liquidator, address borrower, uint seizeTokens) virtual external returns (uint);
```
Comment (arthur-cw) [inline contracts/Cloudwalk/CTokenInterfaces.sol:264]: here you are simply re-adding previous code that was overwritten, right?

**Review by arthur-cw — APPROVED** [review]
Apart from comments in DM and those here, looks good to me :-)

File: contracts/Cloudwalk/CErc20.sol (hunk 20 above)
Comment (igorsenych-cw) [reply contracts/Cloudwalk/CErc20.sol:160]: It's cheaper to initiate a local variable and read from it multiple times than reading from the storage variable multiple times.

File: contracts/Cloudwalk/CTokenInterfaces.sol (hunk 22 above)
Comment (igorsenych-cw) [reply contracts/Cloudwalk/CTokenInterfaces.sol:264]: Yes

**Pull request merged**